File: language/src/pattern.rs
use crate::*;

stored_closure!(pub for<'a> struct Pattern(&'a Value, &Env, &Stack) -> Result<Option<Cow<'a, Value>>>);

impl Primitive for Pattern {}

impl Pattern {
    pub fn any() -> Self {
        Pattern::new(|value, _, _| Ok(Some(Cow::Borrowed(value))))
    }

    pub fn for_trait(r#trait: Trait) -> Self {
        Pattern::new(move |value, env, stack| value.get_trait_if_present(&r#trait, env, stack))
    }

    pub fn is(pattern: Pattern) -> Self {
        Pattern::new(move |value, env, stack| {
            pattern(value, env, stack)?;
            Ok(Some(Cow::Borrowed(value)))
        })
    }
}

File: Language/Sources/Pattern.swift
public struct Pattern {
    public typealias Match = (Value, Env, Stack) throws -> Value?

    public let match: Match

    public init(match: @escaping Match) {
        self.match = match
    }
}

public extension Pattern {
    static let any = Pattern { value, _, _ in value }

    init<T: Primitive>(_ type: T.Type) {
        self.init(trait: Trait(type))
    }

    init(trait: Trait) {
        self.init { value, env, stack in
            try value.getIfPresent(trait: trait, env, stack)
        }
    }

    init(is pattern: Pattern) {
        self.init { value, env, stack in
            _ = try pattern.match(value, env, stack)
            return value
        }
    }
}
